flatsurf/geometry/hyperbolic.py

```diff
@@ -0,0 +1,774 @@
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: HyperbolicPlane().base_ring()
+            Rational Field
+
+        """
+        return self._base_ring
+
+    def infinity(self):
+        r"""
+        Return the point at infinity in the Poincaré half plane model.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: HyperbolicPlane().infinity()
+            ∞
+
+        """
+        return self.projective(1, 0)
+
+    def real(self, r):
+        r"""
+        Return the ideal point ``r`` on the real axis in the Poincaré half
+        plane model.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: HyperbolicPlane().real(-2)
+            -2
+
+        """
+        return self.projective(r, 1)
+
+    def projective(self, p, q):
+        r"""
+        Return the ideal point with projective coordinates ``[p: q]`` in the
+        Poincaré half plane model.
+
+        EXMAPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: H = HyperbolicPlane()
+            sage: H.projective(0, 1)
+            0
+
+            sage: H.projective(-1, 0)
+            ∞
+
+            sage: H.projective(0, 0)
+            Traceback (most recent call last):
+            ...
+            ValueError: one of p and q must not be zero
+
+        """
+        if p == 0 and q == 0:
+            raise ValueError("one of p and q must not be zero")
+
+        if q == 0:
+            return self.point(0, 1, model="klein")
+
+        p = self.base_ring()(p)
+        q = self.base_ring()(q)
+
+        return self.point(p/q, 0, model="half_plane")
+
+    def point(self, x, y, model):
+        r"""
+        Return the point with coordinates (x, y) in the given model.
+
+        When ``model`` is ``"half_plane"``, return the point `x + iy` in the upper half plane.
+
+        When ``model`` is ``"klein"``, return the point (x, y) in the Klein disc.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+            sage: H = HyperbolicPlane()
+
+            sage: H.point(0, 1, model="half_plane")
+            I
+
+            sage: H.point(1, 2, model="half_plane")
+            1 + 2*I
+
+            sage: H.point(0, 1, model="klein")
+            ∞
+
+        """
+        x = self.base_ring()(x)
+        y = self.base_ring()(y)
+
+        if model == "klein":
+            return self.__make_element_class__(HyperbolicPoint)(self, x, y)
+        if model == "half_plane":
+            denominator = 1 + x*x + y*y
+            return self.point(
+                x=2*x / denominator,
+                y=(-1 + x*x + y*y) / denominator,
+                model="klein")
+
+        raise NotImplementedError("unsupported model")
+
+    def half_circle(self, center, radius_squared):
```

Comment:
I am not confortable with the name here as "circle" has another meaning in hyperbolic geometry.

Reply:
Any suggestions? I don't think that anybody would expect a half circle to be a shortcut to get an actual hyperbolic half circle.